manifests/apache.pp

```diff
@@ -1,34 +1,122 @@
 # Configure an Apache vhost
 # @api private
-class pulpcore::apache {
+class pulpcore::apache (
+  String[1] $vhost_priority = '10',
+  Stdlib::Port $http_port = 80,
+  Stdlib::Port $https_port = 443,
+) {
   $api_path = '/pulp/api/v3'
   $api_url = "http://${pulpcore::api_host}:${pulpcore::api_port}${api_path}"
   $content_path = '/pulp/content'
-  $content_url = "http://${pulpcore::content_host}:${pulpcore::content_port}${content_path}"
-
-  if $pulpcore::manage_apache {
-    include apache
-    apache::vhost { 'pulpcore':
-      servername => $pulpcore::servername,
-      port       => 80,
-      priority   => '10',
-      docroot    => $pulpcore::webserver_static_dir,
-      proxy_pass => [
-        {
-          'path'         => $api_path,
-          'url'          => $api_url,
-          'reverse_urls' => [$api_url],
-        },
-        {
-          'path'         => $content_path,
-          'url'          => $content_url,
-          'reverse_urls' => [$content_url],
-        },
+  $content_base_url = "http://${pulpcore::content_host}:${pulpcore::content_port}"
+  $content_url = "${content_base_url}${content_path}"
+
+  $base_directories = [
+    {
+      provider       => 'directory',
+      path           => $pulpcore::webserver_static_dir,
+      options        => ['Indexes','FollowSymLinks'],
+      allow_override => ['None'],
+    },
+    {
+      'path'            => $content_path,
+      'provider'        => 'location',
+      'request_headers' => [
+        'unset X-CLIENT-CERT',
+        'set X-CLIENT-CERT "%{SSL_CLIENT_CERT}s" env=SSL_CLIENT_CERT',
       ],
+    },
+  ]
+
+  if $pulpcore::remote_user_environ_name {
+    $remote_user_environ_header = $pulpcore::remote_user_environ_name.regsubst(/^HTTP_/, '')
+    $api_directories = [
+      {
+        'path'            => $api_path,
+        'provider'        => 'location',
+        'request_headers' => [
+          "unset ${remote_user_environ_header}",
+          "set ${remote_user_environ_header} \"%{SSL_CLIENT_S_DN_CN}s\" env=SSL_CLIENT_S_DN_CN",
+        ],
+      },
+    ]
+  } else {
+    $api_directories = []
+  }
+
+  $proxy_pass_content = {
+    'path' => $content_path,
+    'url'  => $content_url,
+  }
+
+  $proxy_pass_api = {
+    'path' => $api_path,
+    'url'  => $api_url,
+  }
```

Comment:
Previously, these resulting ProxyPass declarations, when used with the Foreman vhost were included from a directory. That include was near the top of the vhost and therefore those ProxyPass rules got declared before the primary ones defined in the Foreman vhost (https://github.com/theforeman/puppet-foreman/blob/master/manifests/config/apache.pp#L110). These changes the ordering such that now the `/pulp` paths are unreachable since their ProxyPass declarations come after the these declarations.

Trouble is, can you force the proxy pass declarations on the vhost via fragment to appear in a particular order? We cannot drop `/pulp !` as since the Foreman's ProxyPass is a catchall on `/` we would not get `/pulp` routes connecting properly either.

Reply:
The concat module uses a priority on every fragment and then alphabetically sorts them to determine the order. Right now this doesn't use priority on fragments, but that would be the way to go.

puppet-foreman should not have to know about what other modules are adding to it. I'd like it if Foreman didn't have to specify `/pulp !` and we rely on inserting fragments before the actual ProxyPass statement.

Reply:
Does this mean there are changes needed here or on puppet-foreman to fix this?

Reply:
Probably not - we use the standard parameters on puppetlabs-apache. That order is probably [here](https://github.com/puppetlabs/puppetlabs-apache/blob/f8af65497f7198eac5e6eedd4bae529f561f2da4/manifests/vhost.pp#L2454-L2485). So using `order => 155` in our fragments would be a good thing to test with.